### ISSUE TITLE
empty attribute values are omitted in rendering

### ISSUE THREAD

**Issue by bergus** [ndmitchell/tagsoup#23]
I'd like to reopen [`renderTags` render empty attributes `<body style="">` as `<body style>`](http://code.google.com/p/ndmitchell/issues/detail?id=267)

I do have some problems with this, as in the legacy html I am processing there are some content-less attributes that were included solely for being there (like `<img alt="" …>`), and I don't want to change anything about that (for now, of course a proper solution is omitting them entirely or filling them with values).

If you don't want to alter the data model (like wrapping the value in `Maybe`), a quick fix might be to distinguish "boolean" attributes which are commonly empty from "text" attributes, so that you can both produce the empty attribute syntax for the former and the empty string value for the latter respectively. This would also work around the problems with the doctype that were mentioned in the google code issue.

**Comment by ndmitchell** [comment]
I don't want to alter the data model at the moment - it seems that this is a bit of a corner case, but altering the data model would impact everyone. Also, in newer HTML standards, you're meant to do things like `selected="selected"` anyway.

So are you after a flag which makes all empty attributes print as `""`? Or on a per-attribute or per-tag basis? And if a per-thing basis, per which thing?

**Comment by bergus** [comment]
I think HTML5 does encourage `<input required …>` and `<option selected>` anyway - and as I have just learned, empty attribute syntax is allowed anywhere [\[1\]](http://stackoverflow.com/questions/32949632/for-which-html5-attributes-is-the-empty-attribute-syntax-allowed). Indeed hardly a need to change the data model, unless you plan to support `∀ html: render (parse html) == html`.

I'm not looking for a flag (I'm not even sure how I would get Hakyll into using it), I thought an educated guess on a per-attribute basis could do.
Attributes like `required`, `selected`, etc. would get the empty attribute syntax (no value), attributes like `alt`, `style` or `href` would get the empty string value syntax. If you want, I think I can compile a complete list for you.

**Comment by ndmitchell** [comment]
Since TagSoup is not HTML specific, but can handle any XML, I'd need to make it parameterisable, much like `optMinimize` is already, and then do it for the specific named attributes. A complete list of attributes to apply to would be a good start.

**Comment by bergus** [comment]
Hm, I'm pretty sure though that empty attribute syntax makes invalid XML, so the default should be to create empty strings anyway :-)

**Comment by ndmitchell** [comment]
Apart from my treating the doctype as XML, which then has the fake no-attribute things, although I certainly agree it's usually invalid. Maybe whether to render empty attributes then depends on the tag and the attribute name? So for XML, tags starting `!` would not have empty quotes, but other ones would.

**Comment by bergus** [comment]
Yeah, if it could depend on the tag name as well that would surely avoid issues with `!DOCTYPE` tags, we wouldn't even need to whitelist `html` and `PUBLIC` attributes categorically.

**Comment by ndmitchell** [comment]
Cool, do you want to put down the initial tag/attribute pairs you think should have this treatment applied to them? Once I have that, I'll try and write the patch (or you might want to give it a go? it's not in sensitive code and should be relatively straightforward - happy to advise if you'd like to take that route)

**Comment by bergus** [comment]
OK, here's an overview. For definition of empty attributes, see
* [HTML4 boolean attributes](http://www.w3.org/TR/html4/intro/sgmltut.html#h-3.3.4.2)
* [HTML5 boolean attributes](http://www.w3.org/TR/html5/infrastructure.html#boolean-attributes) / [WHATWG](https://html.spec.whatwg.org/multipage/infrastructure.html#boolean-attributes)
* [HTML5 empty attribute syntax](http://www.w3.org/TR/html5/syntax.html#attributes-0) / [WHATWG](https://html.spec.whatwg.org/multipage/syntax.html#attributes-2)

I did use the [HTML4](http://www.w3.org/TR/html4/index/attributes.html) and [HTML5](https://html.spec.whatwg.org/multipage/indices.html#attributes-3) attribute indices to compile the following list of boolean attributes and the elements they may appear on:
* `allowfullscreen`: `iframe`
* `async`: `script`
* `autofocus`: `button`, `input`, `keygen`, `select`, `textarea`
* `autoplay`: `audio`, `video`
* `checked`: `menuitem`, `input`
* `controls`: `audio`, `video`
* `default`: `menuitem`, `track`
* `defer`: `script`
* `disabled`: `menuitem`, `fieldset`, `keygen`, `button`, `input`, `optgroup`, `option`, `select`, `textarea`
* `formnovalidate`: `button`, `input`
* `hidden`: all html elements
* `ismap`: `img`, `input`
* `itemscope`: all html elements
* `loop`: `audio`, `video`
* `multiple`: `input`, `select`
* `muted`: `audio`, `video`
* `novalidate`: `form`
* `open`: `details`, `dialog`
* `readonly`: `textarea`, `input`
* `required`: `input`, `select`, `textarea`
* `reversed`: `ol`
* `scoped`: `style`
* `seamless`: `iframe`
* `selected`: `option`
* `sortable`: `table`
* `typemustmatch`: `object`
* `compact`: `dir`, `dl`, `menu`, `ol`, `ul`
* `declare`: `object`
* `nohref`: `area`
* `noresize`: `frame`
* `noshade`: `hr`
* `nowrap`: `td`, `th`

Also as JSON:
```json
{"allowfullscreen":["iframe"],"async":["script"],"autofocus":["button","input","keygen","select","textarea"],"autoplay":["audio","video"],"checked":["menuitem","input"],"controls":["audio","video"],"default":["menuitem","track"],"defer":["script"],"disabled":["menuitem","fieldset","keygen","button","input","optgroup","option","select","textarea"],"formnovalidate":["button","input"],"hidden":"*","ismap":["img","input"],"itemscope":"*","loop":["audio","video"],"multiple":["input","select"],"muted":["audio","video"],"novalidate":["form"],"open":["details","dialog"],"readonly":["textarea","input"],"required":["input","select","textarea"],"reversed":["ol"],"scoped":["style"],"seamless":["iframe"],"selected":["option"],"sortable":["table"],"typemustmatch":["object"],"compact":["dir","dl","menu","ol","ul"],"declare":["object"],"nohref":["area"],"noresize":["frame"],"noshade":["hr"],"nowrap":["td","th"]}
```
and reversed (tag -> attributes):
```json
{"*":["hidden","itemscope"],"iframe":["allowfullscreen","seamless"],"script":["async","defer"],"button":["autofocus","disabled","formnovalidate"],"input":["autofocus","checked","disabled","formnovalidate","ismap","multiple","readonly","required"],"keygen":["autofocus","disabled"],"select":["autofocus","disabled","multiple","required"],"textarea":["autofocus","disabled","readonly","required"],"audio":["autoplay","controls","loop","muted"],"video":["autoplay","controls","loop","muted"],"menuitem":["checked","default","disabled"],"track":["default"],"fieldset":["disabled"],"optgroup":["disabled"],"option":["disabled","selected"],"img":["ismap"],"form":["novalidate"],"details":["open"],"dialog":["open"],"ol":["reversed","compact"],"style":["scoped"],"table":["sortable"],"object":["typemustmatch","declare"],"dir":["compact"],"dl":["compact"],"menu":["compact"],"ul":["compact"],"area":["nohref"],"frame":["noresize"],"hr":["noshade"],"td":["nowrap"],"th":["nowrap"]}
```

**Comment by marrus-sh** [comment]
I would say that there are definitely instances where non-boolean attributes set with the empty string are not the same as omitting the value; namely:

1. + `<img />` gives an image with no provided label
    + `<img alt="" />` gives an image whose label is explicitly an empty string, signalling that the image is purely visual in nature
    + `<img alt="alt" />` gives an image whose label is "alt"

2. Similarly,
    + `<button>Foo</button>` gives a button whose accessible label is "Foo"
    + `<button aria-label="">Foo</button>` gives a button whose accessible label is the empty string
    + `<button aria-label="aria-label">Foo</button>` gives a button whose accessible label is "aria-label"

3. And of course,
    + `<div></div>` has `dataset.foo == undefined`
    + `<div data-foo=""></div>` has `dataset.foo == ""`
    + `<div data-foo="data-foo"></div>` has `data-foo == "data-foo"`

So empty attributes (`<img alt />`) should not be dropped or given content. And `<img alt="" />` is not only much clearer than `<img alt />`, but it is correct Polyglot markup (it is valid XML).

Contrarily, there is no good reason to favour `<div hidden></div>` over `<div hidden=""></div>` aside from saving oneself three characters. So I definitely think the empty-quote syntax is the better choice here, and can be applied in all cases (without need for checking whether a given attribute is boolean).

**Comment by ndmitchell** [comment]
@marrus-sh - sounds pretty convincing. Does that match the current behaviour of tagsoup?

**Comment by marrus-sh** [comment]
as far as i'm aware, current behaviour of tagsoup is to render `<div hidden></div>` rather than `<div hidden=""></div>`, as described in the link in the first comment on this issue

**Comment by ndmitchell** [comment]
OK, happy to take a patch fixing that! I'm not really an active maintainer of this package, so also happy to get a co-maintainer if anyone wants.
